src/array.cr

```diff
@@ -1545,22 +1544,67 @@ class Array(T)
     self
   end
 
+  # Returns `self` with all the elements shifted `n` times.
+  #
+  # ```
+  # a1 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # a2 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  # a3 = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+  #
+  # a1.rotate!
+  # a2.rotate!(1)
+  # a3.rotate!(3)
+  #
+  # a1 # => [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
+  # a2 # => [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
+  # a3 # => [3, 4, 5, 6, 7, 8, 9, 0, 1, 2]
+  # ```
   def rotate!(n = 1)
     return self if size == 0
+    # This is broken into cases for perfromance (#8515)
     n %= size
-    return self if n == 0
-    if n <= size // 2
+
+    if n == 0
+      # ignore
+    elsif n == 1
+      tmp = self[0]
+      @buffer.move_from(@buffer + n, size - n)
+      self[-1] = tmp
+    elsif n == (size - 1)
+      tmp = self[-1]
+      (@buffer + size - n).move_from(@buffer, n)
+      self[0] = tmp
+    elsif n <= SMALL_ARRAY_SIZE // 2 # shift right
```

Comment:
Why `// 2` on this branch and not the left shift branch?

Reply:
shift left and shift right is describing the direction of the algorithm.

Reply:
Yes but why is it `<= SMALL_ARRAY_SIZE` on one branch and `<= SMALL_ARRAY_SIZE // 2` on the other?

Reply:
so this comment might be out of date to the most current commit.  right now it is checking size.
```cr
elsif n <= size // 2
```

But the reason why one is checking against the `SMALL_ARRAY_SIZE` and the other `size` is because one is being optimized.  As for the `//2` and not I dont know.

Reply:
I meant

```
elsif n <= SMALL_ARRAY_SIZE // 2
```

vs

```
elsif size - n <= SMALL_ARRAY_SIZE
```

Reply:
that works it was just not how I was thinking about it.

Reply:
??

This is two excerpts from your PR

Reply:
This conditional has changed.  Do you still have a concern about this?